core/arch/arm/plat-sunxi/psci.c

```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <console.h>
+#include <io.h>
+#include <stdint.h>
+#include <kernel/generic_boot.h>
+#include <kernel/misc.h>
+#include <kernel/panic.h>
+#include <kernel/pm_stubs.h>
+#include <mm/core_mmu.h>
+#include <mm/core_memprot.h>
+#include <mm/tee_pager.h>
+#include <platform_config.h>
+#include <sm/tee_mon.h>
+#include <sm/optee_smc.h>
+#include <sm/psci.h>
+
+#define REG_CPUCFG_RES0             (0x0000)
+#define REG_CPUCFG_CPU_RST(cpu)     (0x0040 + (cpu) * (0x0040))
+#define REG_CPUCFG_GEN_CTRL         (0x0184)
+#define REG_CPUCFG_PRIV0            (0x01a4)
+#define REG_CPUCFG_DBG_CTRL1        (0x01e4)
+#define REG_PRCM_CPU_PWROFF         (0x0100)
+#define REG_PRCM_CPU_PWR_CLAMP(cpu) (0x0140 + (cpu) * (0x0004))
+
+static void cp15_write_cntp_tval(uint32_t);
+static void cp15_write_cntp_ctl(uint32_t);
+static uint32_t cp15_read_cntp_ctl(void);
+static void __mdelay(uint32_t);
+
+int psci_features(uint32_t psci_fid)
+{
+	switch (psci_fid) {
+#ifdef CFG_BOOT_SECONDARY_REQUEST
+	case PSCI_CPU_ON:
+		return 0;
+#endif
+
+	default:
+		return PSCI_RET_NOT_SUPPORTED;
+	}
+}
+
+static void cp15_write_cntp_tval(uint32_t tval)
+{
+	asm volatile ("mcr p15, 0, %0, c14, c2, 0" : : "r" (tval));
+}
+
+static void cp15_write_cntp_ctl(uint32_t val)
+{
+	asm volatile ("mcr p15, 0, %0, c14, c2, 1" : : "r" (val));
+}
+
+static uint32_t cp15_read_cntp_ctl(void)
+{
+	uint32_t val;
+
+	asm volatile ("mrc p15, 0, %0, c14, c2, 1" : "=r" (val));
+
+	return val;
+}
+
+static void __mdelay(uint32_t ms)
```

Comment:
This does not belong in `plat-sunxi` I think. We do have `udelay()` in `<kernel/delay.h>`. I'm not sure if the implementation would be usable here, but even if a different implementation is needed, it should go next to `core/arch/arm/kernel/delay.c` and be selected based on some `CFG_` flag.